File: chap18_gastrointestinal/shoot.f90
use shootcom
implicit none

diff=1000.0d0; r=0.05d0; c0=0.3; P=0.2d0; startbigL=100.0d0; N0=1.0d-1
npts=10000
!littlel=bigL/r; eps=diff/(r*c0*P); littlen=N0/(c0*c0*P)

open(1,file='shoot.dat')

print*,'enter 1 for single solution'
print*,'enter 2 for efflux conc as function of channel length'
read*,ichoice

if (ichoice .eq. 1) nLstep=1
if (ichoice .ne. 1) nLstep=50

do Lstep=1,nLstep
        if (ichoice .eq. 1) bigL=startbigL
	if (ichoice .ne. 1) bigL= startbigL - (Lstep-1.0d0)*(startbigL-10.0d0)/(float(nLstep-1))
        del=bigL/float(npts)

	cinitup=15.0d0
	cinitdown=0.01d0
	c(1)=cinitdown; call doshoot; if (ichoice .eq. 1) print*, shootout
	c(1)=cinitup; call doshoot; if (ichoice .eq. 1) print*,shootout
	
	shootout=10.0d0
	
	do while (dabs(shootout) > 0.001)
	    c(1)=(cinitup+cinitdown)/2.0d0
	    call doshoot
	    if (shootout<0) cinitdown=(cinitup+cinitdown)/2.0d0
	    if (shootout>0) cinitup=(cinitup+cinitdown)/2.0d0
            if (ichoice == 1) then
	        print*,cinitdown,cinitup,shootout
            endif
	end do
	
        dcdx=(c(npts)-c(npts-1))/del
        outos=c0-diff*dcdx/v(npts)
        if (ichoice .eq. 1) then
		do i=1,npts
		    write(1,*) c(i),v(i)
		end do	
		print*,'solute outflow=',c(npts)*v(npts) - diff*dcdx
		print*,'solute inflow=',2.0d0*(bigL/10.0d0)*N0/r
		print*,'outflow osmolarity =', outos
        endif
	if (ichoice .ne. 1) then
                print*, bigL,outos
	        write(1,*) bigL,outos
	endif
end do

end

!-----------------------------------------------
subroutine doshoot
use shootcom
implicit none
double precision :: N

d(1)=0.0d0; v(1)=0.0d0
do i=1,npts-1
    dist=(i-1)*del
    c(i+1) = c(i) + del*d(i)
    d(i+1) = d(i) + del*(1.0d0/diff)*(v(i)*d(i) + c(i)*2.0d0*(P/r)*(c(i)-c0) - 2.0d0*N(dist)/r)
    v(i+1) = v(i) + del*2.0d0*(P/r)*(c(i)-c0)
    if (c(i+1)<0.0) then
        shootout=-100.0d0
        return
    endif
    if (c(i+1)>100.0d0) then
        shootout=100.0
        return
    endif
end do

shootout=c(npts)-c0

return
end

!-----------------------------
function N(dum)
use shootcom
implicit none
double precision :: N,dum

N=0.0d0
if (dum<bigL/10.0d0) N=N0

return
end
!-------------------------------

File: chap18_gastrointestinal/shoot.m

function main

clear all;
global diff r c0 P startbigL N0 npts del bigL

diff=1000.0; r=0.05; c0=0.3; P=0.2; startbigL=100.0; N0=0.3;
npts=10000;

bigL=startbigL;
del=bigL/npts;

cinitup=15.0;
cinitdown=0.01; 
[dist c out]=doshoot(cinitup);
plot(dist,c) 
axis([0 100 0 1.3])
hold on
[dist c out]=doshoot(cinitdown);
plot(dist,c)

for i=1:30
    cinit=(cinitup+cinitdown)/2.0d0;
	[dist c out]= doshoot(cinit);
    plot(dist,c)
	if (out<0) 
        cinitdown=(cinitup+cinitdown)/2.0d0;
    end
	if (out>0) 
        cinitup=(cinitup+cinitdown)/2.0d0;
    end
end
    
hold off

%--------------------------------------

function [dist c out]=doshoot(cinit)
global diff r c0 P startbigL N0 npts del bigL

d(1)=0.0; v(1)=0.0; c(1)=cinit;
for i=1:npts-1
    dist(i+1)=(i-1)*del;
    c(i+1) = c(i) + del*d(i);
    d(i+1) = d(i) + del*(1.0/diff)*(v(i)*d(i) + c(i)*2.0*(P/r)*(c(i)-c0) - 2.0*bigN(dist)/r);
    v(i+1) = v(i) + del*2.0*(P/r)*(c(i)-c0);
    if (c(i+1)<0.0)
        out=-100.0;
        return
    end
    if (c(i+1)>100.0)
        out=100.0;
        return
    end
end

out=c(npts)-c0;

%--------------------------------------

function out=bigN(dum)
global diff r c0 P startbigL N0 npts del bigL

out=0.0;
if (dum<bigL/10.0)
    out=N0;
end
